protos/mongodb/mongodb_parser.go

```diff
@@ -0,0 +1,323 @@
+
+// see http://docs.mongodb.org/meta-driver/latest/legacy/mongodb-wire-protocol/#op-reply
+func opReplyParse(d *decoder, m *MongodbMessage) (bool, bool) {
+	_, err := d.readInt32() // ignore flags for now
+	m.event["cursorId"], err = d.readInt64()
+	m.event["startingFrom"], err = d.readInt32()
+
+	numberReturned, err := d.readInt32()
+	m.event["numberReturned"] = numberReturned
+
+	logp.Debug("mongodb", "Prepare to read %i document from reply", m.event["numberReturned"])
+
+	documents := make([]string, numberReturned)
+	for i := 0; i < numberReturned; i++ {
+		var document bson.M
+		document, err = d.readDocument()
+
+		// Check if the result is actually an error
+		if i == 0 {
+			if mongoError, present := document["$err"]; present {
+				m.error, err = doc2str(mongoError)
+			}
+		}
+
+		documents[i], err = doc2str(document)
+	}
+	m.event["documents"] = documents
+
+	if err != nil {
+		logp.Err("An error occured while parsing OP_REPLY message: %s", err)
+		return false, false
+	}
+	return true, true
+}
+
+func opMsgParse(d *decoder, m *MongodbMessage) (bool, bool) {
+	var err error
+	m.event["message"], err = d.readCStr()
+	if err != nil {
+		logp.Err("An error occured while parsing OP_MSG message: %s", err)
+		return false, false
+	}
+	return true, true
+}
+
+func opUpdateParse(d *decoder, m *MongodbMessage) (bool, bool) {
+	_, err := d.readInt32() // always ZERO, a slot reserved in the protocol for future use
+	m.event["fullCollectionName"], err = d.readCStr()
+	_, err = d.readInt32() // ignore flags for now
+
+	m.event["selector"], err = d.readDocumentStr()
+	m.event["update"], err = d.readDocumentStr()
+
+	if err != nil {
+		logp.Err("An error occured while parsing OP_UPDATE message: %s", err)
+		return false, false
+	}
+
+	return true, true
+}
+
+func opInsertParse(d *decoder, m *MongodbMessage) (bool, bool) {
+	_, err := d.readInt32() // ignore flags for now
+	m.event["fullCollectionName"], err = d.readCStr()
+
+	// TODO parse bson documents
+	// Not too bad if it is not done, as all recent mongodb clients send insert as a command over a query instead of this
+	// Find an old client to generate a pcap with legacy protocol ?
+
+	if err != nil {
+		logp.Err("An error occured while parsing OP_INSERT message: %s", err)
+		return false, false
+	}
+
+	return true, true
+}
+
+func opQueryParse(d *decoder, m *MongodbMessage) (bool, bool) {
+	_, err := d.readInt32() // ignore flags for now
+	fullCollectionName, err := d.readCStr()
+	m.event["fullCollectionName"] = fullCollectionName
+
+	m.event["numberToSkip"], err = d.readInt32()
+	m.event["numberToReturn"], err = d.readInt32()
+
+	query, err := d.readDocument()
+	if d.i < len(d.in) {
+		m.event["returnFieldsSelector"], err = d.readDocumentStr()
+	}
+
+	// Actual method is either a 'find' or a command passing through a query
+	if strings.HasSuffix(fullCollectionName, ".$cmd") {
+		m.method = "otherCommand"
+		for _, command := range UserCommands {
+			if _, present := query[command]; present {
```

Comment:
Here we need to also check that the the value is a string, for example: `update: restaurants` should match, but `update: {}` should not. This is needed for the `updateandmodify` command.